File: examples3d/fountain3.rs
use rapier_testbed3d::TestbedViewer;
use rapier3d::prelude::*;

const MAX_NUMBER_OF_BODIES: usize = 400;

pub async fn run(viewer: &mut TestbedViewer) -> anyhow::Result<()> {
    let mut world = PhysicsWorld::new();

    let rad = 0.5;

    /*
     * Ground
     */
    let ground_size = 40.0;
    let ground_height = 2.1; // 16.0;

    for k in 0..3 {
        let rigid_body =
            RigidBodyBuilder::fixed().translation(Vector::new(0.0, -ground_height - k as f32, 0.0));
        let collider = ColliderBuilder::cuboid(ground_size, ground_height, ground_size);
        let (_handle, _) = world.insert(rigid_body, collider);
    }

    /*
     * Set up the viewer.
     */
    viewer.set_world(&mut world);
    viewer.look_at(Vec3::new(30.0, 4.0, 30.0), Vec3::new(0.0, 1.0, 0.0));

    /*
     * Run the simulation, spawning bodies from the fountain each step and
     * pruning the oldest ones once the cap is reached (was an `add_callback`).
     */
    let mut step_id = 0usize;
    while viewer.render_frame(&mut world).await {
        if viewer.simulating() {
            world.step();
            step_id += 1;

            let rigid_body = RigidBodyBuilder::dynamic().translation(Vector::new(0.0, 10.0, 0.0));
            let handle = world.bodies.insert(rigid_body);
            let collider = match step_id % 3 {
                0 => ColliderBuilder::round_cylinder(rad, rad, rad / 10.0),
                1 => ColliderBuilder::cone(rad, rad),
                _ => ColliderBuilder::cuboid(rad, rad, rad),
            };

            world
                .colliders
                .insert_with_parent(collider, handle, &mut world.bodies);
            viewer.add_body(handle, &world);

            if world.bodies.len() > MAX_NUMBER_OF_BODIES {
                let mut to_remove: Vec<(RigidBodyHandle, Vector)> = world
                    .bodies
                    .iter()
                    .filter(|e| e.1.is_dynamic())
                    .map(|e| (e.0, e.1.translation()))
                    .collect();

                to_remove.sort_by(|a, b| {
                    (a.1.x.abs() + a.1.z.abs())
                        .partial_cmp(&(b.1.x.abs() + b.1.z.abs()))
                        .unwrap()
                        .reverse()
                });

                let num_to_remove = to_remove.len().saturating_sub(MAX_NUMBER_OF_BODIES);
                for (handle, _) in &to_remove[..num_to_remove] {
                    world.bodies.remove(
                        *handle,
                        &mut world.islands,
                        &mut world.colliders,
                        &mut world.impulse_joints,
                        &mut world.multibody_joints,
                        true,
                    );
                    viewer.remove_body(*handle);
                }
            }
        }
    }

    Ok(())
}

File: bindings/python/rapier-testbed/rapier_testbed/examples3/fountain3.py
"""Port of examples3d/fountain3.rs.

Continuously spawns dynamic primitives that fall on three stacked ground
slabs. When the dynamic body count exceeds a cap, the farthest ones are
removed each frame.
"""
from __future__ import annotations

import rapier3d as rp
from .._registry import register

CATEGORY = "Collisions"
NAME = "Fountain"

MAX_NUMBER_OF_BODIES = 400


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    rad = 0.5
    ground_size = 40.0
    ground_height = 2.1

    for k in range(3):
        body = rp.RigidBody.fixed().translation((0.0, -ground_height - float(k), 0.0))
        h = bodies.insert(body)
        colliders.insert_with_parent(
            rp.Collider.cuboid(ground_size, ground_height, ground_size), h, bodies
        )

    def _cb(tb) -> None:
        rb = rp.RigidBody.dynamic().translation((0.0, 10.0, 0.0))
        h = tb.bodies.insert(rb)
        step = tb._step_count
        kind = step % 3
        if kind == 0:
            col = rp.Collider.round_cylinder(rad, rad, rad / 10.0)
        elif kind == 1:
            col = rp.Collider.cone(rad, rad)
        else:
            col = rp.Collider.cuboid(rad, rad, rad)
        tb.colliders.insert_with_parent(col, h, tb.bodies)

        if len(tb.bodies) > MAX_NUMBER_OF_BODIES:
            dyn = [
                (hh, body)
                for hh, body in tb.bodies
                if body.is_dynamic
            ]
            dyn.sort(
                key=lambda pair: abs(pair[1].translation.x) + abs(pair[1].translation.z),
                reverse=True,
            )
            num_remove = max(0, len(dyn) - MAX_NUMBER_OF_BODIES)
            for handle, _ in dyn[:num_remove]:
                tb.bodies.remove(
                    handle,
                    tb._islands,
                    tb.colliders,
                    tb.impulse_joints,
                    tb.multibody_joints,
                    True,
                )

    testbed.add_callback(_cb)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((30.0, 4.0, 30.0), (0.0, 1.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
